### ISSUE TITLE
Why dropping 24.3?

### ISSUE THREAD

**Issue by syl20bnr** [Bruce-Connor/paradox#35]
I wonder what is the motivation to drop 24.3 ?
This version is still shipped with Ubuntu 14.4 LTS and I guess some other distros.

I will have to maintain 2 versions of paradox which is not ideal. Looking at the code, the compatibility cost in LoC seemed not very high.

**Comment by Bruce-Connor** [comment]
> This version is still shipped with Ubuntu 14.4 LTS and I guess some other distros.

I understand that, and I'm not happy to have dropped it. I try to support at least 24.1 in all of my packages. However, in Paradox's case, this was turning into more of a chore than usual (see below). 

Note, however:

1. The stable version still supports 24.3, so I haven't removed anything from them yet.
2. I'm trying to think of a way to still offer version 1.5 once 2.0 becomes stable. So they'll always have paradox.
3. I describe below what would need to be done so that I can continue developing for 24.3.

> Looking at the code, the compatibility cost in LoC seemed not very high.

The code was about 20% larger. It's not huge, but it's not trivial to maintain either. But the amount of LoC is not the real reason. (see below)

> I wonder what is the motivation to drop 24.3 ? 

The way packages are internally described changed significantly in 24.4 (or package.el version 1.0.1). They changed from manually created data structures to elisp classes. This means I had to do a bit of extra work at almost every turn. When you add the fact that I haven't managed to write good tests for paradox yet, that also means I had to manually check every new feature worked with 24.3.

Ultimately, this extra brain power and effort required really added up, and I had stopped implementing cool new things because of this extra effort it would take. Paradox is something I do for fun, and updating it had turned into a bit of a chore.

> I will have to maintain 2 versions of paradox which is not ideal.

What are you referring to here?

-----

# What could be done?

If paradox had its own (unified) interface for package descriptions, and if it had a few more tests, there would be no extra effort involved in supporting 24.3 and we could maintain support. That was the original intention behind the `paradox-compat.el` file, but my time ran low before it ever got finished.

**Comment by Bruce-Connor** [comment]
Now that I think about it. It might be possible to distribute `package.el` version 1.0.1 through one of the repositories (melpa or GNU Elpa). In this case we could support 24.1 with no extra effort.

I'll ask about this at emacs-devel

**Comment by syl20bnr** [comment]
Thank you for the detailed answer.

> What are you referring to here?

Paradox is distributed via MELPA for users using 24.4+. For users using 24.3 it is distributed as an extension which means that I copied the last version of Paradox compatible with 24.3 in the `extensions` folder of the spacemacs configuration layer.

For now the configuration of Paradox remains the same no matter the version in use. But may be in the future I will have to consider to different configurations, for instance if a new cool feature is offered in the 24.4 version only.

So basically I'm forced to maintain the configuration for 2 versions of Paradox if I want that spacemacs users can continue to enjoy Paradox, which should not be my concern in the first place. With this decision you put the burden further on the Emacs kit maintainer's shoulders. 

But I understand your decision and agree with it and I'm fine with the current situation, Paradox is a great package!

If it becomes messy to support the two versions I will try to get some time to help with the refactoring.

**Comment by Bruce-Connor** [comment]
> But I understand your decision and agree with it and I'm fine with the current situation, Paradox is a great package!

Thanks for the comprehension, I do not take this decision lightly. In the end, if I hadn't dropped support there wouldn't be any more cool new features.

>So basically I'm forced to maintain the configuration for 2 versions of Paradox if I want that spacemacs users can continue to enjoy Paradox, which should not be my concern in the first place. 

I promise I'll try to design new releases so that configuration is fully backwards compatible. That is: if you set some variables for the 2.0 version, but the user actually has 1.5 installed, no harm should come of having configured these variables.

For the moment, this is still true. The only new variable is the `post-execute-functions`, and using `add-hook` will not throw any errors if the hook variable doesn't exist.

**Comment by syl20bnr** [comment]
> I promise I'll try to design new releases so that configuration is fully backwards compatible. That is: if you set some variables for the 2.0 version, but the user actually has 1.5 installed, no harm should come of having configured these variables.

That would be great !

Count on me to at least report any issue users may encounter in 24.3.

Thank you.
